From 9170f0b2e56207fd21ed1848069cec77d7fe90a4 Mon Sep 17 00:00:00 2001
From: Xiaofei Ma <xiaofei@motorola.com>
Date: Fri, 2 Oct 2015 15:20:35 -0500
Subject: [PATCH 459/959] IKQCOMSD1-2948: arm/dt: 8952: Support Fingerprint
 Sensor

- Remove i2c-8 which is used for spi.
- Add spi pins configuration
- Add fpc dts

Change-Id: I1fc68b99ee26afd2a765711d280d54563e03fe7c
Signed-off-by: Xiaofei Ma <xiaofei@motorola.com>
Reviewed-on: http://gerrit.mot.com/788089
SME-Granted: SME Approvals Granted
SLTApproved: Slta Waiver <sltawvr@motorola.com>
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Thomas Merrell <thomasmerrell@motorola.com>
Reviewed-by: Igor Kovalenko <igork@motorola.com>
Reviewed-by: Elliot Stewart <elliot@motorola.com>
Reviewed-by: Dmitry Ornatsky <ornatsky@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
---
 .../devicetree/bindings/input/misc/fpc1020.txt     |  31 +++++
 .../devicetree/bindings/vendor-prefixes.txt        |   1 +
 arch/arm/boot/dts/qcom/msm8952-athene-p0.dts       |   2 +-
 arch/arm/boot/dts/qcom/msm8952-athene_16mp-p1.dts  |   2 +
 arch/arm/boot/dts/qcom/msm8952-moto-common.dtsi    |   3 +-
 arch/arm/boot/dts/qcom/msm8952-moto-fps.dtsi       | 140 +++++++++++++++++++++
 6 files changed, 177 insertions(+), 2 deletions(-)
 create mode 100644 Documentation/devicetree/bindings/input/misc/fpc1020.txt
 create mode 100644 arch/arm/boot/dts/qcom/msm8952-moto-fps.dtsi

diff --git a/Documentation/devicetree/bindings/input/misc/fpc1020.txt b/Documentation/devicetree/bindings/input/misc/fpc1020.txt
new file mode 100644
index 00000000000..6389e7539ec
--- /dev/null
+++ b/Documentation/devicetree/bindings/input/misc/fpc1020.txt
@@ -0,0 +1,31 @@
+FPC fingerprint sensor driver.
+
+Required properties:
+
+ - compatible		: Should be "fpc,fpc1020".
+ - reg			: spi slave address of the device.
+ - pinctrl-names	: Pinctrl configuration names of this sensor driver.
+				Should be "default".
+ - pinctrl-0		: The pinctrl node corresponding to "default",
+				Should be <&fpc1020_default>.
+ - interrupt-parent	: Parent of interrupt.
+ - interrupts		: Accelerometer interrupt to indicate new data ready or events.
+ - vdd_io-supply	: Digital IO power supply needed for sensor.
+
+Example:
+&spi_7 { /* BLSP2 QUP4 */
+	fpc_fpc1020@0{
+		compatible = "fpc,fpc1020";
+		reg = <0>;
+		interrupts = <42 0x0>;
+		fpc,gpio_irq = <42>;
+		fpc,gpio_reset = <47>;
+		fpc,gpio_cs0 = <98>;
+		pinctrl-names = "default", "active";
+		pinctrl-0 = <&fps_rst_default &fps_int_default>;
+		pinctrl-1 = <&fps_rst_active &fps_int_active>;
+		fpc,enable-on-boot;
+		fpc,txout_boost_enable;
+		vdd_io-supply  = <&pm8950_l6>;
+	};
+};
diff --git a/Documentation/devicetree/bindings/vendor-prefixes.txt b/Documentation/devicetree/bindings/vendor-prefixes.txt
index 28bd3c34c8c..6afc3d954a3 100644
--- a/Documentation/devicetree/bindings/vendor-prefixes.txt
+++ b/Documentation/devicetree/bindings/vendor-prefixes.txt
@@ -28,6 +28,7 @@ epson	Seiko Epson Corp.
 est	ESTeem Wireless Modems
 fairchild	Fairchild Semiconductor International, Inc.
 focaltech Focaltech systems
+fpc	Fingerprint Cards
 fsl	Freescale Semiconductor
 GEFanuc	GE Fanuc Intelligent Platforms Embedded Systems, Inc.
 goodix	Goodix. Ltd.
diff --git a/arch/arm/boot/dts/qcom/msm8952-athene-p0.dts b/arch/arm/boot/dts/qcom/msm8952-athene-p0.dts
index 782ff39460d..cab1f9c2620 100644
--- a/arch/arm/boot/dts/qcom/msm8952-athene-p0.dts
+++ b/arch/arm/boot/dts/qcom/msm8952-athene-p0.dts
@@ -15,7 +15,7 @@
 
 #include "msm8952-athene.dtsi"
 #include "dsi-panel-mot-tdi-546-1080p-cmd.dtsi"
-
+#include "msm8952-moto-fps.dtsi"
 / {
 	qcom,board-id = <0x41 0x8000>, <0x41 0x800d>;
 };
diff --git a/arch/arm/boot/dts/qcom/msm8952-athene_16mp-p1.dts b/arch/arm/boot/dts/qcom/msm8952-athene_16mp-p1.dts
index 9be186455e4..972a8927dc5 100644
--- a/arch/arm/boot/dts/qcom/msm8952-athene_16mp-p1.dts
+++ b/arch/arm/boot/dts/qcom/msm8952-athene_16mp-p1.dts
@@ -14,6 +14,8 @@
 /dts-v1/;
 
 #include "msm8952-athene.dtsi"
+#include "dsi-panel-mot-tdi-546-1080p-cmd.dtsi"
+#include "msm8952-moto-fps.dtsi"
 
 / {
 	model = "Athene_16MP";
diff --git a/arch/arm/boot/dts/qcom/msm8952-moto-common.dtsi b/arch/arm/boot/dts/qcom/msm8952-moto-common.dtsi
index 2caca8ae964..8416ff64a33 100644
--- a/arch/arm/boot/dts/qcom/msm8952-moto-common.dtsi
+++ b/arch/arm/boot/dts/qcom/msm8952-moto-common.dtsi
@@ -22,6 +22,7 @@
 		serial0 = &blsp1_uart2;
 		spi4 = &spi_4;
 		/delete-property/ i2c5;
+		/delete-property/ i2c8;
 		/delete-property/ spi0;
 	};
 
@@ -359,7 +360,7 @@
 &soc {
 	/delete-node/ spi@78B5000;
 	/delete-node/ i2c@7af5000;
-
+	/delete-node/ i2c@7af8000;
 	ramoops {
 		compatible = "ramoops";
 		status = "ok";
diff --git a/arch/arm/boot/dts/qcom/msm8952-moto-fps.dtsi b/arch/arm/boot/dts/qcom/msm8952-moto-fps.dtsi
new file mode 100644
index 00000000000..f381cb9308d
--- /dev/null
+++ b/arch/arm/boot/dts/qcom/msm8952-moto-fps.dtsi
@@ -0,0 +1,140 @@
+/* Copyright (c) 2015, Motorola Mobility Inc. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 and
+ * only version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+/ {
+	aliases {
+		spi7 = &spi_7;
+	};
+};
+
+&tlmm_pinmux {
+	fps_int {
+		qcom,pins = <&gp 42>;
+		qcom,num-grp-pins = <1>;
+		qcom,pin-func = <0>;
+		label = "fps-int";
+
+		fps_int_default: default {
+			output-low;
+		};
+
+		fps_int_active: active {
+			drive-strength = <10>;
+			bias-pull-up; /* pull-up */
+		};
+	};
+
+	fps_rst_n {
+		qcom,pins = <&gp 47>;
+		qcom,num-grp-pins = <1>;
+		qcom,pin-func = <0>;
+		label = "fps_rst";
+
+		fps_rst_default: default {
+			bias-disable;
+		};
+
+		fps_rst_active: active {
+			drive-strength = <10>;
+			bias-pull-up; /* pull-up */
+		};
+	};
+
+	spi7_active {
+		/* BLSP8 MOSI, MISO */
+		qcom,pins = <&gp 96>, <&gp 97>;
+		qcom,num-grp-pins = <2>;
+		qcom,pin-func = <1>;
+		label = "spi7-active";
+		/* active state */
+		spi7_default: spi7_default {
+			drive-strength = <12>; /* 12 MA */
+			bias-disable; /* No PULL */
+		};
+	};
+
+	spi7_suspend {
+		/* BLSP8 MOSI, MISO */
+		qcom,pins = <&gp 96>, <&gp 97>;
+		qcom,num-grp-pins = <2>;
+		qcom,pin-func = <1>;
+		label = "spi7-suspend";
+		/* suspended state */
+		spi7_sleep: spi7_sleep {
+			drive-strength = <2>; /* 2 MA */
+			bias-pull-down; /* PULL Down */
+		};
+	};
+
+	spi7_cs0_active {
+		/* BLSP8 CS, CLK */
+		qcom,pins = <&gp 98>, <&gp 99>;
+		qcom,num-grp-pins = <2>;
+		qcom,pin-func = <0>;
+		label = "spi7-cs0-active";
+		spi7_cs0_active: cs0_active {
+			drive-strength = <2>;
+			bias-disable;
+		};
+	};
+
+	spi7_cs0_suspend {
+		/* BLSP8 CS, CLK */
+		qcom,pins = <&gp 98>, <&gp 99>;
+		qcom,num-grp-pins = <2>;
+		qcom,pin-func = <0>;
+		label = "spi7-cs0-suspend";
+		spi7_cs0_sleep: cs0_sleep {
+			drive-strength = <2>;
+			bias-pull-up;
+		};
+	};
+};
+
+&soc {
+	spi_7: spi@7AF8000 {
+		compatible = "qcom,spi-qup-v2";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		reg-names = "spi_physical";
+		reg = <0x7AF8000 0x600>;
+		interrupt-names = "spi_irq";
+		interrupts = <0 302 0>;
+		spi-max-frequency = <19200000>;
+		pinctrl-names = "spi_default", "spi_sleep";
+		pinctrl-0 = <&spi7_default &spi7_cs0_active>;
+		pinctrl-1 = <&spi7_sleep &spi7_cs0_sleep>;
+		qcom,infinite-mode = <0>;
+		qcom,use-pinctrl;
+		qcom,master-id = <84>;
+
+		fpc_fpc1020@0{
+			compatible = "fpc,fpc1020";
+			reg = <0>;
+			interrupts = <42 0x0>;
+			spi-max-frequency = <4800000>;
+			fpc,gpio_irq = <&msm_gpio 42 0>;
+			fpc,gpio_rst = <&msm_gpio 47 0>;
+			pinctrl-names = "default", "active";
+			pinctrl-0 = <&fps_rst_default &fps_int_default>;
+			pinctrl-1 = <&fps_rst_active &fps_int_active>;
+			fpc,enable-on-boot;
+			fpc,txout_boost_enable;
+			vcc_spi-supply = <&pm8950_l6>;
+			vcc_ana-supply = <&pm8950_l6>;
+			vcc_io-supply  = <&pm8950_l6>;
+			clock-names = "iface_clk", "core_clk";
+			clocks = <&clock_gcc clk_gcc_blsp2_ahb_clk>,
+				<&clock_gcc clk_gcc_blsp2_qup4_spi_apps_clk>;
+		};
+	};
+};
-- 
2.11.0

